File: app/src/main/java/com/amaury/pointage/v2/engine/EmployerMobilityContributionV2.kt
package com.amaury.pointage.v2.engine

import java.time.YearMonth

/**
 * Versement mobilité employeur : aucune applicabilité ni aucun taux n'est déduit automatiquement.
 * Les règles doivent être confirmées pour une période et conserver une source vérifiable.
 */
object EmployerMobilityContributionV2 {
    enum class Status { APPLICABLE, NOT_APPLICABLE }

    data class Record(
        val id: String,
        val status: Status,
        /** Taux décimal confirmé : 2,50 % = 0,025. Null lorsque non applicable. */
        val rate: Double?,
        val effectiveFrom: YearMonth,
        val effectiveTo: YearMonth? = null,
        val source: String
    )

    data class Snapshot(
        val applicable: Boolean?,
        val rate: Double?,
        val source: String?,
        val reliable: Boolean,
        val warnings: List<String>
    )

    data class Result(
        val employerAmount: Double?,
        val complete: Boolean,
        val warnings: List<String>
    )

    fun resolve(records: List<Record>, period: YearMonth): Snapshot {
        val malformed = records.filter { record ->
            record.effectiveTo?.let { it < record.effectiveFrom } == true ||
                record.source.isBlank() ||
                (record.status == Status.APPLICABLE && (record.rate == null || !record.rate.isFinite() || record.rate < 0.0 || record.rate > 1.0))
        }
        if (malformed.isNotEmpty()) {
            return Snapshot(
                applicable = null,
                rate = null,
                source = null,
                reliable = false,
                warnings = listOf("Versement mobilité employeur : une règle enregistrée est incomplète ou incohérente ; aucun taux n'est appliqué.")
            )
        }

        val active = records.filter { record ->
            period >= record.effectiveFrom && (record.effectiveTo == null || period <= record.effectiveTo)
        }
        if (active.isEmpty()) {
            return Snapshot(
                applicable = null,
                rate = null,
                source = null,
                reliable = false,
                warnings = listOf("Versement mobilité employeur : applicabilité et taux à confirmer pour ${period.monthValue.toString().padStart(2, '0')}/${period.year}.")
            )
        }
        if (active.size > 1) {
            return Snapshot(
                applicable = null,
                rate = null,
                source = null,
                reliable = false,
                warnings = listOf("Versement mobilité employeur : plusieurs règles se chevauchent sur la période ; calcul patronal bloqué.")
            )
        }

        val selected = active.single()
        return when (selected.status) {
            Status.NOT_APPLICABLE -> Snapshot(
                applicable = false,
                rate = 0.0,
                source = selected.source,
                reliable = true,
                warnings = emptyList()
            )
            Status.APPLICABLE -> Snapshot(
                applicable = true,
                rate = selected.rate,
                source = selected.source,
                reliable = true,
                warnings = emptyList()
            )
        }
    }

    fun calculate(grossSocial: Double, rate: Double?): Result {
        if (rate == null) {
            return Result(
                employerAmount = null,
                complete = false,
                warnings = listOf("Versement mobilité employeur : taux applicable à confirmer ; coût employeur incomplet.")
            )
        }
        if (!rate.isFinite() || rate < 0.0 || rate > 1.0) {
            return Result(
                employerAmount = null,
                complete = false,
                warnings = listOf("Versement mobilité employeur : taux invalide ; aucun montant patronal n'est calculé.")
            )
        }
        val base = grossSocial.coerceAtLeast(0.0)
        return Result(
            employerAmount = base * rate,
            complete = true,
            warnings = emptyList()
        )
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/EmployerMobilityContributionV2.swift
import Foundation

/// Versement mobilité employeur : aucune applicabilité ni aucun taux n'est déduit automatiquement.
/// Les règles doivent être confirmées pour une période et conserver une source vérifiable.
enum EmployerMobilityContributionV2 {
    enum Status: Equatable {
        case applicable
        case notApplicable
    }

    struct Period: Equatable, Comparable {
        let year: Int
        let month: Int

        var isValid: Bool { year > 0 && (1...12).contains(month) }

        static func < (lhs: Period, rhs: Period) -> Bool {
            lhs.year == rhs.year ? lhs.month < rhs.month : lhs.year < rhs.year
        }

        var label: String { String(format: "%02d/%04d", month, year) }
    }

    struct Record: Equatable {
        let id: String
        let status: Status
        let rate: Double?
        let effectiveFrom: Period
        let effectiveTo: Period?
        let source: String
    }

    struct Snapshot: Equatable {
        let applicable: Bool?
        let rate: Double?
        let source: String?
        let reliable: Bool
        let warnings: [String]
    }

    struct Result: Equatable {
        let employerAmount: Double?
        let complete: Bool
        let warnings: [String]
    }

    static func resolve(records: [Record], period: Period) -> Snapshot {
        guard period.isValid else {
            return unresolved("Versement mobilité employeur : période invalide ; aucun taux n'est appliqué.")
        }

        let malformed = records.filter { record in
            record.source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                !record.effectiveFrom.isValid ||
                (record.effectiveTo?.isValid == false) ||
                (record.effectiveTo.map { $0 < record.effectiveFrom } ?? false) ||
                (record.status == .applicable && (record.rate.map { !validRate($0) } ?? true))
        }
        guard malformed.isEmpty else {
            return unresolved("Versement mobilité employeur : une règle enregistrée est incomplète ou incohérente ; aucun taux n'est appliqué.")
        }

        let active = records.filter { record in
            period >= record.effectiveFrom && (record.effectiveTo == nil || period <= record.effectiveTo!)
        }
        guard !active.isEmpty else {
            return unresolved("Versement mobilité employeur : applicabilité et taux à confirmer pour \(period.label).")
        }
        guard active.count == 1, let selected = active.first else {
            return unresolved("Versement mobilité employeur : plusieurs règles se chevauchent sur la période ; calcul patronal bloqué.")
        }

        switch selected.status {
        case .notApplicable:
            return Snapshot(
                applicable: false,
                rate: 0,
                source: selected.source,
                reliable: true,
                warnings: []
            )
        case .applicable:
            return Snapshot(
                applicable: true,
                rate: selected.rate,
                source: selected.source,
                reliable: true,
                warnings: []
            )
        }
    }

    static func calculate(grossSocial: Double, rate: Double?) -> Result {
        guard let rate else {
            return incomplete("Versement mobilité employeur : taux applicable à confirmer ; coût employeur incomplet.")
        }
        guard validRate(rate) else {
            return incomplete("Versement mobilité employeur : taux invalide ; aucun montant patronal n'est calculé.")
        }
        guard grossSocial.isFinite, grossSocial >= 0 else {
            return incomplete("Versement mobilité employeur : assiette brute sociale invalide ; aucun montant patronal n'est calculé.")
        }

        return Result(
            employerAmount: grossSocial * rate,
            complete: true,
            warnings: []
        )
    }

    private static func unresolved(_ warning: String) -> Snapshot {
        Snapshot(
            applicable: nil,
            rate: nil,
            source: nil,
            reliable: false,
            warnings: [warning]
        )
    }

    private static func incomplete(_ warning: String) -> Result {
        Result(employerAmount: nil, complete: false, warnings: [warning])
    }

    private static func validRate(_ rate: Double) -> Bool {
        rate.isFinite && rate >= 0 && rate <= 1
    }
}
